File: gemma-web/src/main/java/ubic/gemma/web/taglib/DatabaseEntryTag.java
/*
 * The Gemma project
 *
 * Copyright (c) 2006 University of British Columbia
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package ubic.gemma.web.taglib;

import org.apache.commons.lang3.StringUtils;
import org.springframework.web.servlet.tags.form.TagWriter;
import ubic.gemma.core.loader.util.ExternalDatabaseUtils;
import ubic.gemma.model.common.description.DatabaseEntry;
import ubic.gemma.model.common.description.DatabaseEntryValueObject;
import ubic.gemma.web.assets.StaticAssetResolver;
import ubic.gemma.web.util.ExternalDatabaseWebUtils;

import javax.annotation.Nullable;
import javax.servlet.jsp.JspException;
import javax.servlet.jsp.tagext.Tag;

/**
 * @author keshav
 */
public class DatabaseEntryTag extends AbstractHtmlElementTag {

    private transient StaticAssetResolver staticAssetResolver;

    @Nullable
    private DatabaseEntryValueObject databaseEntry;

    @Override
    public int doStartTagInternal() throws JspException {
        if ( staticAssetResolver == null ) {
            staticAssetResolver = getRequestContext().getWebApplicationContext().getBean( StaticAssetResolver.class );
        }

        TagWriter tagWriter = new TagWriter( pageContext );

        if ( this.databaseEntry == null ) {
            tagWriter.startTag( "i" );
            tagWriter.appendValue( "No accession available" );
            tagWriter.endTag();
            return SKIP_BODY;
        }

        tagWriter.startTag( "span" );
        writeOptionalAttributes( tagWriter );

        boolean hasLabel = StringUtils.isNotBlank( databaseEntry.getLabel() );

        if ( hasLabel ) {
            tagWriter.appendValue( htmlEscape( databaseEntry.getLabel() ) );
            tagWriter.appendValue( " " );
        }

        if ( databaseEntry.getExternalDatabase() != null ) {
            String externalUri = ExternalDatabaseUtils.getUri( databaseEntry );
            String databaseLogo = ExternalDatabaseWebUtils.getLogo( databaseEntry.getExternalDatabase() );
            if ( externalUri != null && isHttpUrl( externalUri ) ) {
                if ( databaseLogo != null ) {
                    tagWriter.startTag( "a" );
                    tagWriter.writeAttribute( "href", externalUri );
                    tagWriter.writeAttribute( "target", "_blank" );
                    tagWriter.writeAttribute( "rel", "noopener noreferrer" );
                    writeDatabaseLogo( databaseEntry.getExternalDatabase().getName(), databaseLogo, tagWriter );
                    tagWriter.endTag(); // </a>
                } else {
                    if ( hasLabel ) {
                        tagWriter.appendValue( "(" );
                    }
                    tagWriter.startTag( "a" );
                    tagWriter.writeAttribute( "href", externalUri );
                    tagWriter.writeAttribute( "target", "_blank" );
                    tagWriter.writeAttribute( "rel", "noopener noreferrer" );
                    tagWriter.appendValue( htmlEscape( databaseEntry.getExternalDatabase().getName() ) );
                    tagWriter.appendValue( " " );
                    tagWriter.startTag( "i" );
                    tagWriter.writeAttribute( "class", "fa fa-external-link" );
                    tagWriter.endTag( true );
                    tagWriter.endTag(); // </a>
                    if ( hasLabel ) {
                        tagWriter.appendValue( ")" );
                    }
                }
            } else if ( databaseLogo != null ) {
                writeDatabaseLogo( databaseEntry.getExternalDatabase().getName(), databaseLogo, tagWriter );
            } else {
                if ( hasLabel ) {
                    tagWriter.appendValue( "(" );
                }
                tagWriter.appendValue( "(" + htmlEscape( databaseEntry.getExternalDatabase().getName() ) + ")" );
                if ( hasLabel ) {
                    tagWriter.appendValue( ")" );
                }
            }
        }

        tagWriter.endTag();

        return Tag.SKIP_BODY;
    }

    private boolean isHttpUrl( String uri ) {
        return uri.startsWith( "http://" ) || uri.startsWith( "https://" );
    }

    private void writeDatabaseLogo( String databaseName, String databaseLogo, TagWriter tagWriter ) throws JspException {
        tagWriter.startTag( "img" );
        tagWriter.writeAttribute( "src", staticAssetResolver.resolveUrl( databaseLogo ) );
        tagWriter.writeAttribute( "height", "16" );
        tagWriter.writeAttribute( "alt", htmlEscape( databaseName ) + " logo" );
        tagWriter.endTag(); // </img>
    }

    public void setDatabaseEntry( @Nullable DatabaseEntry databaseEntry ) {
        if ( databaseEntry == null ) {
            // if it is a user-owned data set.
            this.databaseEntry = null;
            return;
        }
        this.databaseEntry = new DatabaseEntryValueObject( databaseEntry );
    }

    public void setDatabaseEntryValueObject( DatabaseEntryValueObject databaseEntry ) {
        this.databaseEntry = databaseEntry;
    }
}

File: gemma-web/src/main/webapp/scripts/api/entities/common/DatabaseEntryTag.js
import ExternalDatabaseUtils from "./ExternalDatabaseUtils";

let htmlEncode = Ext.util.Format.htmlEncode;

function isHttpUrl( uri ) {
   return uri.startsWith( 'http://' ) || uri.startsWith( 'https://' );
}

/**
 * Simple component for rendering a database entry.
 *
 * This should be kept as much as possible in sync with the DatabaseEntryTag.java.
 * @author poirigui
 */
export default class DatabaseEntryTag {
   /**
    *
    * @param {DatabaseEntryValueObject} databaseEntry
    */
   constructor( databaseEntry ) {
      this.databaseEntry = databaseEntry;
   }

   render() {
      if ( !this.databaseEntry ) {
         return "<i>No accession available</i>";
      }
      let edMeta = ExternalDatabaseUtils.externalDatabases
         .find( ed => ed.name === this.databaseEntry.externalDatabase.name );
      let s;
      if ( edMeta ) {
         if ( this.databaseEntry.label ) {
            s = htmlEncode( this.databaseEntry.label ) + ' ';
         } else {
            s = '';
         }
         let externalDatabaseLogo = '<img src="' + edMeta.logo + '" height="16" alt="' + htmlEncode( edMeta.name ) + ' logo"/>';
         if ( this.databaseEntry.uri !== null && isHttpUrl( this.databaseEntry.uri ) ) {
            s += '<a target="_blank" rel="noopener noreferrer" href="' + this.databaseEntry.uri + '">' + externalDatabaseLogo + '</a>';
         } else if ( this.databaseEntry.externalDatabase.uri !== null && isHttpUrl( this.databaseEntry.externalDatabase.uri ) ) {
            s += '<a target="_blank" rel="noopener noreferrer" href="' + this.databaseEntry.externalDatabase.uri + '">' + externalDatabaseLogo + '</a>';
         } else {
            // no link available
            s += externalDatabaseLogo;
         }
      } else {
         let externalDatabaseLinkHtml = htmlEncode( this.databaseEntry.externalDatabase.name ) + ' <i class="fa fa-external-link"></i>';
         if ( this.databaseEntry.uri !== null ) {
            if ( this.databaseEntry.label ) {
               s = htmlEncode( this.databaseEntry.label ) + ' (<a target="_blank" rel="noopener noreferrer" href="' + this.databaseEntry.uri + '">' + externalDatabaseLinkHtml + '</a>)';
            } else {
               s = '<a target="_blank" rel="noopener noreferrer" href="' + this.databaseEntry.uri + '">' + externalDatabaseLinkHtml + '</a>';
            }
         } else {
            // no link available
            if ( this.databaseEntry.label ) {
               s = htmlEncode( this.databaseEntry.label ) + ' (' + externalDatabaseLinkHtml + "')";
            } else {
               s = externalDatabaseLinkHtml;
            }
         }
      }
      return "<span>" + s + "</span>";
   }
}